boto/endpoints.py

```diff
@@ -0,0 +1,415 @@
+        # A mapping can be provided that handles the mapping between
+        # "service names" and endpoint prefixes.
+        'awslambda': 'lambda',
+        'cloudwatch': 'monitoring',
+        'ses': 'email',
+        'ec2containerservice': 'ecs',
+        'configservice': 'config',
+    }
+
+    def __init__(self, endpoint_data, boto_endpoint_data=None,
+                 service_rename_map=None):
+        """
+        :type endpoint_data: dict
+        :param endpoint_data: Regions and endpoints data in the same format
+            as is used by botocore / boto3.
+
+        :type boto_endpoint_data: dict
+        :param boto_endpoint_data: Regions and endpoints data in the legacy
+            boto2 format. This data takes precedence over any data found in
+            `endpoint_data`.
+
+        :type service_rename_map: dict
+        :param service_rename_map: A mapping of boto2 service name to
+            endpoint prefix.
+        """
+        super(BotoEndpointResolver, self).__init__(endpoint_data)
+        self._boto_endpoint_data = boto_endpoint_data
+        if service_rename_map is None:
+            service_rename_map = self.SERVICE_RENAMES
+        # Mapping of boto2 service name to endpoint prefix
+        self._endpoint_prefix_map = service_rename_map
+        # Mapping of endpoint prefix to boto2 service name
+        self._service_name_map = dict(
+            (v, k) for k, v in service_rename_map.items())
+
+    def get_available_endpoints(self, service_name, partition_name='aws',
+                                allow_non_regional=False):
+        endpoint_prefix = self._endpoint_prefix(service_name)
+        endpoints = super(BotoEndpointResolver, self).get_available_endpoints(
+            endpoint_prefix, partition_name, allow_non_regional)
+        boto_endpoints = self._get_available_boto_endpoints(service_name)
+        return list(set(endpoints + boto_endpoints))
+
+    def get_all_available_regions(self, service_name):
+        """Retrieve every region across partitions for a service."""
+        regions = set()
+        endpoint_prefix = self._endpoint_prefix(service_name)
+
+        # Get every region for every partition in the new endpoint format
+        for partition_name in self.get_available_partitions():
+            if self.is_global_service(service_name, partition_name):
+                # Global services are available in every region in the
+                # partition in which they are considered global.
+                partition = self._get_partition_data(partition_name)
+                regions.update(partition['regions'].keys())
+            else:
+                regions.update(
+                    super(BotoEndpointResolver, self).get_available_endpoints(
+                        endpoint_prefix, partition_name
+                    )
+                )
+
+        # boto2 endpoint format has no partitions, so load them all.
+        boto_endpoints = self._get_available_boto_endpoints(service_name)
```

Comment:
Hmm that makes sense.

I wonder if that begs the question: should the override be in the endpoint resolver logic? Instead should it be an override in the static endpoint builder? I think it would make sense from an interface perspective of the static builder if the interface was:
* It will use the default heuristics to generate a static list
* Then a user can provide an additional set of data to merge over whatever the default heuristics generate

It would make the code much cleaner as you would also be dealing with the same data format.

Additionally, I am not sure if the ``BotoEndpointResolver` necessarily need to have that override logic. It is just going to be a fall back if nothing matches from the statically built list when connection to a region. Right?